keps/sig-api-machinery/5366-graceful-leader-transition/README.md

```diff
@@ -0,0 +1,940 @@
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+<!--
+This section is incredibly important for producing high-quality, user-focused
+documentation such as release notes or a development roadmap. It should be
+possible to collect this information before implementation begins, in order to
+avoid requiring implementors to split their attention between writing release
+notes and implementing the feature itself. KEP editors and SIG Docs
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+This proposal outlines a plan to modify the leader election mechanism for key
+Kubernetes components (kube-scheduler, kube-controller-manager,
+cloud-controller-manager). The goal is to enable these components to gracefully
+release the leader lock and transition back to a follower state without a full
+process restart. This change will be introduced behind a new feature gate.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+Many high-availability (HA) Kubernetes components, including kube-scheduler and
+kube-controller-manager, rely on the leader election library in client-go. The
+current implementation mandates that when a component loses its leader lock, it
+must shut down immediately. This is typically handled by calling
+`klog.FlushAndExit()` in the `OnStoppedLeading` callback.
+
+When the leadership is lost, the component shuts down and waits for the kubelet
+to detect that the component is unhealthy and restart it. This has several
+significant drawbacks:
+
+- High Overhead: Restarting a component process incurs unnecessary computational
+  overhead and increases latency during a leadership transition
+- No Graceful Shutdown: The immediate call to `os.Exit()` prevents any graaceful
+  shutdown or cleanup operations.
+- Limited Scalability and Design Flexibility: The current "all-or-nothing"
+  lease management, where a process holds a single lease for all its
+  controllers, prevents a single process from managing multiple leases
+  independently. This forces a scale-up model rather than a scale-out model
+  and severely restricts dynamic controller management.
```

Comment:
The goals don't address separating controllers.  I suggest removing this for a later potential change.